### ISSUE TITLE
A potential mistake in PS2 1(d)

### ISSUE THREAD

**Issue by qiang951128743** [zyxue/stanford-cs229#3]
Hi, I noticed that you wrote in [1(d) of problemset2](https://github.com/zyxue/stanford-cs229/blob/master/Problem-set-2/1-Logistic-regression-training-stability/ans.ipynb) that w in SVM is forced to have norm 1. I believe that's not the case. In the final optimization problem of svm, the norm is  included in the objective function, but there are no such  constraint that the norm of w has to be 1.

**Comment by zyxue** [comment]
Thank you for your question. Please check Page 6 of http://cs229.stanford.edu/notes/cs229-notes3.pdf. Without constraining ||W||=1 in SVM, just by scaling the linear coefficients, you would achieve as large as your want a margin.  

>In the final optimization problem of svm, the norm is included in the objective function, but there are no such constraint that the norm of w has to be 1.

If I understand correctly, during the formalization process of the final optimization problem, the constraint of ||W|| should have already been taken into consideration. If you have a reference, please  post it and I will confirm.

**Comment by qiang951128743** [comment]
Em... Yes in the note,during the development of svm, there is a constrait ||w||=1. However, the constraint was meant to make the margin wx+b the same as the geometric margin, and our goal is acutally to minimize the geometric margin. In fact, please look at the final primal problem of svm and ||w||^2 is in the objective function. If we demand ||w|| to be 1, then the objective function is constant, making the optimization problem trivial and equivalent to finding a feasible solution, which is not very likely the case. If you see any  built-in classifier that outputs ||w|| that has norm 1, I believe it first solves the final primal problem without considering  ||w||=1 and then scaled it because scaling results in the same hyperplane.  I don't have any reference and I doubt any materical would mention the problem. The reasoning is all based on my knowledge in convex optimization. 
I think it's great that you posted these solutions, which really helped me a lot. I also dectected a few other   mistakes  and I am willing to talk to you about it if you are interested.

**Comment by zyxue** [comment]
...That sounds abstract and hard to follow. I may not have as much experience in convex optimization. Could you please suggest an answer the problem of 1(d) of problemset2, and I'll try to look into it.

Of course, if you see additional errors, please be encouraged to create more issues. That's one of my purposes for posting the answers, to learn.

**Comment by qiang951128743** [comment]
Em, yes in the note , during the development of svm, there was a constraint demanding ||w|| to be 1. However, the constraint was meant to make the margin the same as the geometric margin. And our goal is to minimize the geometric margin.
I don't quite get it when you say "just by scaling the linear coefficients, you would achieve as large as your want a margin", isn't the objective function in the final problem "minimizing" 1/2||w||^2? 
I actually don't have any reference, all the reasoning is based on my knowledge in convex optimization. Believe me, if there is not constraint ||w||=1 in the final optimization problem, the resulting optimal solution doesn't necessarily satisfy ||w||=1. If any bulit-in svm classifier outputs ||w|| with norm 1, I bilieve it  
first solves the problem and then scales w and b with ||w||, because scaling won't change the resulting hyperplane.
I think it's great you posted these solutions and I really benefited a lot. I actually also found a few other issues and we can discuss them if you are interested!

**Comment by qiang951128743** [comment]
I'll write it shortly. Logistic regression doestn't converge in the case where the training points are perfectly seperatable because in that case, if we write down the likelyhood function, the fuction has no maximum(infinity). And because of that, the gradient ascent procedure just keeps climbing to infinity.
SVM is different, SVM loves the perfect seperatable training set because it aims at finding a hyperplane to seperate them. Not perfectly seperatable training points makes the naive SVM invalid and we have to introduce soft-margin to deal with the issue. I am sure you know what I am talking about.

**Comment by zyxue** [comment]
I agree with you that in LR, the likelihood will keep increasing to be infinitely close to 1, so doesn't converge. But that

>SVM is different, SVM loves the perfect seperatable training set because it aims at finding a hyperplane to separate them.

doesn't explain why SVM will do fine. In short, in the case of a perfectly separable training set, both LR and SVM (with linear kernel) look for a hyperplane, but with different goals in mind. 

1. LR searches for a hyperplane that maximizes the likelihood, while 
1. SVM looks for a hyperplane that maximizes the margin.

I recall that the margin is a function of ||W||, so by scaling ||W||, you could get as big a margin as you want. That's why ||W|| needs constrained to 1. I missed some details here, and need to go through the note more closely to work out the details again, maybe later.

**Comment by haroldmei** [comment]
Great discussion! Thanks @qiang951128743. By the way @zyxue has another issue opening for an answer, could you also go have a look? I raised that question and obviously it is still not answered. :)
I am looking forward to an answer of 2.1.(d) as well.

**Comment by haroldmei** [comment]
By the way @zyxue I don't think ps3.3.(b) is finished:
https://github.com/zyxue/stanford-cs229/blob/master/Problem-set-3/3-EM-application-paper-review-modeling.ipynb

At least the update rule shouldn't be dependent on latent distribution, which means you can only use x^(pr) and \mu_p, \nu_r, \delta_p and \tau_r on the right side of the update rule.

Here is another answer, but I also don't think it is complete, but it at least heads the right direction. I've posted him a question as well.
https://github.com/stallmanifold/cs229-machine-learning-stanford-fall-2016/blob/master/homeworks/ps4_key.pdf

Cheers

**Comment by qiang951128743** [comment]
Sure you can get as big a margin as you want. But our goal is not to maximize the margin, in fact, in the final problem, the objective function is to minimize ||w||! It's my mistake that saying the likelyhood function goes to infinity, it actually converges to 1. Em..., please read the svm note again and try to understand the optimization problem, mathematically, not intuitively.

**Comment by zyxue** [comment]
I've updated the [answer for PS2 1(d)](https://github.com/zyxue/stanford-cs229/blob/master/Problem-set-2/1-Logistic-regression-training-stability/ans.ipynb), please let me know if it's still not clear. Special note to the last paragraph,

* you could either fix ||w|| and try to maximize the functional margin
* or you could fix the functional margin and minimize ||w||.

So they are equivalent. I thought the intuition really helps in understanding SVM: to maximize the geometric margin, which won't be affected by scaling of ||w||.

**Comment by zyxue** [comment]
Hi @qiang951128743, do you have question for my updated answer? If not, I am closing this issue.
